vpr/src/place/timing_place.cpp

```diff
@@ -90,16 +73,170 @@ void PlacerCriticalities::update_criticalities(const SetupTimingInfo* timing_inf
          * criticality by taking it to some power, crit_exponent (between 1 and 8 by default). */
         timing_place_crit_[clb_net][pin_index_in_net] = pow(clb_pin_crit, crit_exponent);
     }
+
+    /* Criticalities updated. In sync with timing info.   */
+    /* Can be incrementally updated on the next iteration */
+    recompute_required = false;
+}
+
+/**
+ * @brief Collect the cluster pins which need to be updated based on the latest timing
+ *        analysis so that incremental updates to criticalities can be performed.
+ *
+ * Note we use the set of pins reported by the *timing_info* as having modified
+ * criticality, rather than those marked as modified by the timing analyzer.
+ *
+ * Since timing_info uses shifted/relaxed criticality (which depends on max required
+ * time and worst case slacks), additional nodes may be modified when updating the
+ * atom pin criticalities.
+ */
+
+void PlacerCriticalities::incr_update_criticalities(const SetupTimingInfo* timing_info) {
+    cluster_pins_with_modified_criticality_.clear();
+
+    for (AtomPinId atom_pin : timing_info->pins_with_modified_setup_criticality()) {
```

Comment:
This code and the code to find the pins_with_modified_setup_slack is almost the same, except one uses the timing_info->pins_with_modified_setup_criticality() and the other uses timing_info->pins_with_modified_setup_slack().
1. Do the timing_info->pins_with_modified_setup_criticality() and pins_with_modified_setup_slack() differ?  Why?  I expect they are close to the same; could we just use whichever one is the bigger set (maybe criticality, due to slack shifting) to track both slack and criticality pins that need updates?  Would simplify the code and likely make it faster if we are using both criticalities and slacks, if indeed the pin update lists are usually the same (less work and less memory footprint).

Reply:
If we can just use the bigger of pins_with_modified_criticality_ or pins_with_modified_setup_slack, some data members and some duplicated code can be removed.

Reply:
I don't recommend doing this since they are pre-computed (in ```slack_evaluation.cpp```) and are not necessarily the same. The efforts that go into merging these two sets are not worth it in my opinion, and it might make the code less readable. I agree with your feeling that a lot of codes are duplicated, but I don't have a good way of resolving the issue without sacrificing code readability.